### PR TITLE
File ingest sample for CF

### LINKED ISSUE #61: File ingest sample for CF
Use-case: File -> processing -> some storage.

### LINKED ISSUE #62: Make the SFTP -> DB work in CF
child of spring-cloud/spring-cloud-dataflow-samples#61

### PR DESCRIPTION
**Pull request by chrisjs**
Resolves #61
Resolves #62

### CODE REVIEW

**Comment by chrisjs** [comment]
This sample is probably not mergeable yet as the stream apps are still bringing down spring cloud connector version 2.0.1 which causes a Redis failure. The required version is 2.0.2.

For testing purposes I have a patched JAR created at:https://github.com/chrisjs/scdf-test-stuff/raw/master/sftp-source-rabbit-2.0.0.BUILD-SNAPSHOT.jar - this would need to be used in the `app register --name sftp --type source ...` command for the time being.

Also, until this PR is merged, the pre-compiled demo batch artifact won't be resolvable as it lives in the repo tree. The same artifact can be found at: https://github.com/chrisjs/scdf-test-stuff/raw/master/ingest-sftp-cf-1.0.0.jar - this would be used as the `--batch-resource-uri` parameter to the `stream create` command

**Comment by markpollack** [comment]
Looks like connectors are updated now 
```
$ jar -tvf sftp-source-rabbit-2.0.0.BUILD-20180607.165846-30.jar | grep conn
1254200 Fri May 11 19:05:26 EDT 2018 BOOT-INF/lib/spring-cloud-cloudfoundry-connector-2.0.2.RELEASE.jar
 43139 Fri May 11 19:04:52 EDT 2018 BOOT-INF/lib/spring-cloud-connectors-core-2.0.2.RELEASE.jar
 85227 Fri May 11 19:05:30 EDT 2018 BOOT-INF/lib/spring-cloud-spring-service-connector-2.0.2.RELEASE.jar
```

File: src/main/asciidoc/batch/file-ingest-sftp/main.adoc [hunk 1]
```diff
@@ -226,11 +226,289 @@ These files should be deleted from `/tmp/remote-files` prior to deleting them fr
...
+    SPRING_CLOUD_DEPLOYER_CLOUDFOUNDRY_DOMAIN: YOUR_CF_DOMAIN
+    SPRING_CLOUD_DEPLOYER_CLOUDFOUNDRY_USERNAME: YOUR_CF_USER
+    SPRING_CLOUD_DEPLOYER_CLOUDFOUNDRY_PASSWORD: YOUR_CF_PASSWORD
+    SPRING_CLOUD_DEPLOYER_CLOUDFOUNDRY_STREAM_SERVICES: rabbit
+    SPRING_CLOUD_DEPLOYER_CLOUDFOUNDRY_TASK_SERVICES: mysql
+    SPRING_CLOUD_DEPLOYER_CLOUDFOUNDRY_SKIP_SSL_VALIDATION: true
+    SPRING_APPLICATION_JSON: '{"maven": { "remote-repositories": { "repo1": { "url": "https://repo.spring.io/libs-release"}, "repo2": { "url": "https://repo.spring.io/libs-snapshot"}, "repo3": { "url": "https://repo.spring.io/libs-milestone"} } } }'
+services:
+  - mysql
+  - rabbit
```
Comment (markpollack) [inline src/main/asciidoc/batch/file-ingest-sftp/main.adoc]: the SCDF server itself does not need to be bound to rabbit.

**Comment by chrisjs** [comment]
just took a spin through this and looks like things are working with the latest SNAPSHOT's as indicated in the docs. @sobychacko mentioned they are now bringing down boot 2.0.3.RELEASE so that may be why - also there will be a GA release this week. so maybe keep the SNAPSHOT's for now and bump up to the GA when released.

File: src/main/asciidoc/batch/file-ingest-sftp/main.adoc [hunk 2]
```diff
@@ -226,11 +226,289 @@ These files should be deleted from `/tmp/remote-files` prior to deleting them fr
...
++
+Now a stream needs to be created that will poll the SFTP server, launching the batch job when new files arrive.
++
+Create the stream:
++
+NOTE: You must replace `--username=user`, `--password=pass` and `--host=1.1.1.1` below to their respective values. The `--username=` and `--password=` parameters are the credentials for your remote SFTP user. The `--batch-resource-uri=` parameter is the path to the batch artifact to use. In this Stream definition, the published sample batch artifact JAR is used. If you would like to use a custom built artifact, replace this value with the artifact location.
++
+[source,console]
+----
+dataflow:>stream create --name inboundSftp --definition "sftp --username=user --password=pass --host=1.1.1.1 --allow-unknown-keys=true --task-launcher-output=true --remote-dir=/tmp/remote-files --batch-resource-uri=https://raw.githubusercontent.com/spring-cloud/spring-cloud-dataflow-samples/master/batch/file-ingest-sftp-cf/artifacts/ingest-sftp-cf-1.0.0.jar --local-file-path-job-parameter-value=/tmp/ | task-launcher-cloudfoundry --spring.cloud.deployer.cloudfoundry.services=mysql"
```
Comment (markpollack) [inline src/main/asciidoc/batch/file-ingest-sftp/main.adoc]: The default polling interval is 1s, which is too short and also creates a very busy log file.  Add `fixed-delay=60` or something...

File: src/main/asciidoc/batch/file-ingest-sftp/main.adoc [hunk 3]
```diff
@@ -226,11 +226,289 @@ These files should be deleted from `/tmp/remote-files` prior to deleting them fr
...
+
+. Register the the SFTP source and the Task Launcher Cloud Foundry sink
++
+With the Spring Cloud Data Flow server running, the `SFTP` source and `task-launcher-cloudfoundry` sink needs to be registered. The `SFTP` source application will do the work of polling for new files and when received, it sends a message to the `task-launcher-cloudfoundry` to launch the batch job for that file.
++
+In the Spring Cloud Data Flow shell:
++
+[source,console,options=nowrap]
+----
+dataflow:>app register --name sftp --type source --uri maven://org.springframework.cloud.stream.app:sftp-source-rabbit:2.0.0.BUILD-SNAPSHOT
```
Comment (markpollack) [inline src/main/asciidoc/batch/file-ingest-sftp/main.adoc]: The 2.0 builds are a bit unstable.  `maven://org.springframework.cloud.stream.app:sftp-source-rabbit:1.3.1.RELEASE
` looks to work just fine.

File: src/main/asciidoc/batch/file-ingest-sftp/main.adoc [hunk 4]
```diff
@@ -226,11 +226,289 @@ These files should be deleted from `/tmp/remote-files` prior to deleting them fr
...
+Successfully registered application 'source:sftp'
+dataflow:>app register --name task-launcher-cloudfoundry --type sink --uri maven://org.springframework.cloud.stream.app:task-launcher-cloudfoundry-sink-rabbit:2.0.0.BUILD-SNAPSHOT
```
Comment (markpollack) [inline src/main/asciidoc/batch/file-ingest-sftp/main.adoc]: Similar for `maven://org.springframework.cloud.stream.app:task-launcher-local-sink-rabbit:1.3.1.RELEASE`.

File: batch/file-ingest-sftp-cf/data/people.csv [hunk 5]
```diff
@@ -0,0 +1,3 @@
+Jane,Doe
+John,Doe
+Joe,Doe
```
Comment (markpollack) [inline batch/file-ingest-sftp-cf/data/people.csv]: Add \n

Comment (chrisjs) [reply batch/file-ingest-sftp-cf/data/people.csv]: this was intentional - IIRC adding an empty line causes an exception as the line mapper would have no values to split. there is a configuration option `linesToSkip` but applies to the top of the file https://docs.spring.io/spring-batch/trunk/reference/html/readersAndWriters.html#flatFileItemReader

**Comment by markpollack** [comment]
merged as bdc9d1848dc91e037ad8e2eed7cd67f33437641c

**Pull request closed without merging**